cdap-data-fabric/src/main/java/co/cask/cdap/data2/metrics/HBaseDatasetMetricsReporter.java

```diff
@@ -94,25 +98,37 @@ protected final ScheduledExecutorService executor() {
     return executor;
   }
 
-  private void reportHBaseStats() throws IOException {
-    Map<String, HBaseTableUtil.TableStats> tableStats = hBaseTableUtil.getTableStats(hAdmin);
+  private void reportHBaseStats() throws IOException, DatasetManagementException {
+    Map<TableId, HBaseTableUtil.TableStats> tableStats = hBaseTableUtil.getTableStats(conf, hAdmin);
     if (tableStats.size() > 0) {
       report(tableStats);
     }
   }
 
-  private void report(Map<String, HBaseTableUtil.TableStats> datasetStat) {
-    // we use "0" as runId: it is required by metrics system to provide something at this point
-    for (Map.Entry<String, HBaseTableUtil.TableStats> statEntry : datasetStat.entrySet()) {
-      Id.DatasetInstance datasetInstance = userDsNamespace.fromNamespaced(statEntry.getKey());
-      if (datasetInstance == null) {
-        // not a user dataset
+  private void report(Map<TableId, HBaseTableUtil.TableStats> tableStats) {
+    for (Map.Entry<TableId, HBaseTableUtil.TableStats> statEntry : tableStats.entrySet()) {
+      String namespace = statEntry.getKey().getNamespace().getId();
+      // emit metrics for only user datasets, namespaces in system and
+      // tableNames that doesn't start with user are ignored
+      if (namespace.equals(Constants.SYSTEM_NAMESPACE)) {
         continue;
       }
-      MetricsCollector collector =
-        metricsService.getCollector(ImmutableMap.of(Constants.Metrics.Tag.DATASET, datasetInstance.getId()));
-      // legacy format: dataset name is in the tag. See DatasetInstantiator for more details
-      collector.increment("dataset.size.mb", statEntry.getValue().getTotalSizeMB());
+      String tableName = statEntry.getKey().getTableName();
+      try {
+        Collection<DatasetSpecificationSummary> instances = dsFramework.getInstances(statEntry.getKey().getNamespace());
+        for (DatasetSpecificationSummary spec : instances) {
+          String dsName = spec.getName();
+          if (tableName.startsWith(dsName)) {
```

Comment:
I think a 'startsWith' is not good enough. You could match tables to datasets that don't really belong.

A better way is to check the spec (and all of its embedded specifications) to check equality with the table name, right?